front/lib/stores/appStore.tsx

```diff
@@ -0,0 +1,186 @@
+  timestamp: null,
+  agentConfigurationsCache: [],
+  actions: {
+    setUserId: () => {},
+    getAgentConfigurationsCache: () => [],
+    setAgentConfigurationsCache: () => {},
+    resetAgentConfigurationsCache: () => {},
+  },
+};
+
+export const createGlobalStateStore = () => {
+  return createStore<AppStoreState>()(
+    persist(
+      devtools((set, get) => ({
+        ...initialState,
+        agentConfigurationsCache: [],
+        actions: {
+          setUserId: (userId: string) =>
+            set({ userId }, undefined, "setUserId"),
+          getAgentConfigurationsCache: ({ userId, workspaceId }) => {
+            if (!userId) {
+              return [];
+            }
+
+            const currentCachedUserId = get().userId;
+            const currentCachedWorkspaceId = get().workspaceId;
+
+            if (
+              (currentCachedUserId && userId !== get().userId) ||
+              (currentCachedWorkspaceId &&
+                workspaceId !== currentCachedWorkspaceId)
+            ) {
+              get().actions.resetAgentConfigurationsCache({
+                userId,
+                workspaceId,
+              });
+              return [];
+            }
+
+            const currentTimestamp = Date.now();
+            const lastSavedTimestamp = get().timestamp;
+
+            // if it's more than 7 days old (TBD), let it expire.
+            if (
+              lastSavedTimestamp &&
+              currentTimestamp - lastSavedTimestamp > 1000 * 60 * 60 * 24 * 7
+            ) {
+              get().actions.resetAgentConfigurationsCache({
+                userId,
+                workspaceId,
+              });
+              return [];
+            }
+
+            return get().agentConfigurationsCache;
+          },
+          setAgentConfigurationsCache: ({
+            userId,
+            workspaceId,
+            agentConfigurationsCache,
+          }) =>
+            set(
+              () => {
+                if (!userId) {
+                  return initialState;
+                }
+
+                const currentTimestamp = Date.now();
+
+                return {
+                  userId,
+                  workspaceId,
+                  agentConfigurationsCache,
+                  timestamp: currentTimestamp,
+                };
+              },
+              undefined,
+              "setAgentConfigurationsCache"
+            ),
+          resetAgentConfigurationsCache: ({ userId, workspaceId }) =>
+            set(
+              {
+                userId,
+                workspaceId,
+                agentConfigurationsCache: [],
+                timestamp: null,
+              },
+              undefined,
+              "resetAgentConfigurationsCache"
+            ),
+          resetStore: () => {
+            set(initialState); // We should reset on logout?
+          },
+        },
+      })),
+      {
+        name: "global-store",
+        storage: createJSONStorage(() => indexedDBStorage),
```

Comment:
This saves the state with JSON.stringify(), it's not gonna work with Map and Set and we need to customize storage engine: https://zustand.docs.pmnd.rs/integrations/persisting-store-data#how-do-i-use-it-with-map-and-set